yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnTimelineUtils.scala

```diff
@@ -0,0 +1,720 @@
+    if (value != null ) {
+      Some(value.toString)
+    } else {
+      None
+    }
+  }
+
+  /**
+   * Lookup a numeric field in the `otherInfo` section of a [[TimelineEntity]],
+   * fall back to `defval` if the field is absent or cannot be parsed
+   * @param en entity
+   * @param name field name
+   * @param defval default value; default is 0L
+   * @return the value
+   */
+  def numberField(en: TimelineEntity, name: String, defval:Long = 0L) : Number = {
+    try {
+      fieldOption(en, name) match {
+        case Some(n: Number) => n
+        case _ => defval
+      }
+    } catch {
+      case NonFatal(e) => defval
+    }
+  }
+
+  /**
+   * Take a sequence of timeline events and return an ordered list of spark events.
+   * Important: this reverses the input in the process
+   * @param events event sequence
+   * @return spark event sequence
+   */
+  def asSparkEvents(events: Seq[TimelineEvent]): Seq[SparkListenerEvent] = {
+    events.reverse.map { event =>
+      toSparkEvent(event)
+    }
+  }
+
+  /**
+   * Build date for display in status messages
+   * @param timestamp time in milliseconds post-Epoch
+   * @param unset string to use if timestamp == 0
+   * @return a string for messages
+   */
+  def humanDateCurrentTZ(timestamp: Long, unset: String) : String = {
+    if (timestamp == 0) {
+      unset
+    } else {
+      val dateFormatter = DateFormat.getDateTimeInstance(DateFormat.DEFAULT,
+                                                          DateFormat.LONG)
+      dateFormatter.format(timestamp)
+    }
+  }
+
+  /**
+   * Short formatted time
+   * @param timestamp time in milliseconds post-Epoch
+   * @param unset string to use if timestamp == 0
+   * @return a string for messages
+   */
+  def timeShort(timestamp: Long, unset: String) : String = {
+    if (timestamp == 0) {
+      unset
+    } else {
+      val dateFormatter = DateFormat.getTimeInstance(DateFormat.SHORT)
+      dateFormatter.format(timestamp)
+    }
+  }
+
+  /**
+   * Generate the timeline entity ID from the application and attempt ID.
+   * This is required to be unique across all entities in the timeline server
+   * @param yarnAppId yarn application ID as passed in during creation
+   * @param yarnAttemptId YARN attempt ID as passed in during creation
+   * @param sparkAppId application ID in the application start event
+   * @param attemptId attempt ID in the application start event.
+   */
+  def buildEntityId(yarnAppId: ApplicationId,
+      yarnAttemptId: Option[ApplicationAttemptId],
+      sparkAppId: Option[String],
+      attemptId: Option[String]): String = {
+    yarnAttemptId match {
+      case Some(aid) => aid.toString
+      case None => yarnAppId.toString
+    }
+  }
+
+  /**
+   * Generate the application ID for use in entity fields from the application and attempt ID.
+   * @param yarnAppId yarn application ID as passed in during creation
+   * @param sparkAppId application ID as submitted in the application start event
```

Comment:
This is all related to to the move to app-attempt logic; the choice of which ID to use is implemented entirely inside this code, rather than in tests or the rest of the application; it let me switch between appId, yarnApp id and the one which came in with the spark attempt.

Now that things are fixed and working with yarn-app-id -> entity ID, I'm happy to pull all this out and go direct. Shall I?